Web Extensions: Lets extensions search bookmarks by title and/or URL using Bookmarks API

akutira-umich opened the pull request:
#### fd411fd47c593392ef742cf80a67a054afe168aa
<pre>
Web Extensions: Lets extensions search bookmarks by title and/or URL using Bookmarks API
<a href="https://rdar.apple.com/152505372">rdar://152505372</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297250">https://bugs.webkit.org/show_bug.cgi?id=297250</a>

Reviewed by Brian Weinstein.

Implements the WebProcess and UIProcess for the bookmarks.search() function.
Includes testing as well. Parses the query parameter by each word and locates it in
the list of bookmarks from the bookmarksForExtensionContext delegate function.
For title and url, the types have to match exactly for it be added to the returned array.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::WebExtensionContext::isBookmarksMessageAllowed):
(WebKit::containsIgnoringASCIICase):
(WebKit::normalizeURLStringForComparison):
(WebKit::WebExtensionContext::bookmarksGetRecent):
(WebKit::WebExtensionContext::bookmarksSearch):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::search):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPISearch)):

Canonical link: <a href="https://commits.webkit.org/298686@main">https://commits.webkit.org/298686@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac099db61f76286ad5b5ac1672ff3a9e58bb8cb6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35893 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122289 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66792 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36587 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44481 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88299 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42815 "Found 20 new test failures: fast/canvas/webgl/context-release-upon-reload.html fast/dom/HTMLObjectElement/updateWidget-crash.html fast/forms/checkValidity-001.html fast/forms/form-get-multipart2.html fast/forms/listbox-padding-clip-selected.html fast/html/adjacent-html-context-element.html fast/html/find-selects-draggable-element.html fast/visual-viewport/zoomed-scroll-to-anchor-in-position-fixed.html fast/workers/stress-js-execution.html js/arrowfunction-bind.html ... (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29185 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104284 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68710 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28286 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22393 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65970 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98572 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22539 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125438 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43126 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32381 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97013 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43491 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96798 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24648 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42084 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19976 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39073 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43013 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48605 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42480 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45815 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44184 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -135,6 +135,23 @@
+    return text.containsIgnoringASCIICase(substring);
+}
+
+static WTF::String normalizeURLForComparison(const WTF::String& url)
```
Nit: `normalizeURLStringForComparison`?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -415,6 +432,98 @@
 }
```
Nit: While you're here, you could include an extra blank line.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -895,6 +895,96 @@ void configureMoveBookmarksDelegate(TestWebExtensionManager *manager)
+
+        @"let results9 = await browser.bookmarks.search({ title: 'Dog', url: 'http://animals.com/dog' });",
+        @"browser.test.assertEq(1, results9.length, 'S9: Should find 1 bookmark for title Dog and url animals.com');",
+        @"browser.test.assertEq(bm1.id, results9[0].id, 'S9: Result should be bm1');",
```
Nit: We could do some matching with `https` vs `http`.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -895,6 +895,96 @@ void configureMoveBookmarksDelegate(TestWebExtensionManager *manager)
+
+        @"let results10 = await browser.bookmarks.search({ query: 'animal', title: 'Wild Animal' });",
+        @"browser.test.assertEq(1, results10.length, 'S10: Should find 1 bookmark for query animal and title Wild Animal');",
+        @"browser.test.assertEq(bm6.id, results10[0].id, 'S10: Result should be bm6');",
```
It would be neat to have a test where `query` match the URL and title match the title.

Maybe:
`@"let results10 = await browser.bookmarks.search({ query: 'nature', title: 'Wild Animal' });",`

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -895,6 +895,96 @@ void configureMoveBookmarksDelegate(TestWebExtensionManager *manager)
+        @"browser.test.assertEq(13, results11.length, 'S11: Should return all 10 items for empty query');",
+
+        @"let results12 = await browser.bookmarks.search({ query: 'bookmark A' });",
+        @"browser.test.log(`Starting bookmark test right before...: ${JSON.stringify(results12)}`);",
```
What does this logging indicate? Did you need it for debugging purposes?

akutira-umich replied:
Sorry the logging was necessary then but not anymore.

b-weinstein commented:
This looks nice! A few more tests I'd like to see added but the code changes look very good.

akutira-umich committed:
Web Extensions: Lets extensions search bookmarks by title and/or URL using Bookmarks API
rdar://152505372
https://bugs.webkit.org/show_bug.cgi?id=297250

Reviewed by Brian Weinstein.

Implements the WebProcess and UIProcess for the bookmarks.search() function.
Includes testing as well. Parses the query parameter by each word and locates it in
the list of bookmarks from the bookmarksForExtensionContext delegate function.
For title and url, the types have to match exactly for it be added to the returned array.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::WebExtensionContext::isBookmarksMessageAllowed):
(WebKit::containsIgnoringASCIICase):
(WebKit::normalizeURLStringForComparison):
(WebKit::WebExtensionContext::bookmarksGetRecent):
(WebKit::WebExtensionContext::bookmarksSearch):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::search):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPISearch)):

Canonical link: https://commits.webkit.org/298686@main
